Python: Finalise change notes for 1.24. (#3315)
Semmle/ql: change-notes/1.24/analysis-python.md

## General improvements

- Support for Django version 2.x and 3.x

- Taint tracking now correctly tracks taint in destructuring assignments. For example, if `tainted_list` is a list of tainted tainted elements, then
    ```python
    head, *tail = tainted_list
    ```
    will result in `tail` being tainted with the same taint as `tainted_list`, and `head` being tainted with the taint of the elements of `tainted_list`.

- A large number of libraries and queries have been moved to the new `Value` API, which should result in more precise results.

> - The `Value` API has been extended in various ways:

Reviewer: In documentation, we try to avoid using the term "API" to talk about the CodeQL libraries since it's open to misinterpretation. Would it make sense to talk about the `Value` interface or library instead?

Author: Oh, whoops! We ended up using API the [last time around](https://github.com/Semmle/ql/blob/master/change-notes/1.23/analysis-python.md), so I figured I'd do the same this time. Using "interface" is probably better, though.

> - The `Value` API has been extended in various ways:

Author: ```suggestion
- The `Value` interface has been extended in various ways:
```

   - A new `StringValue` class has been added, for tracking string literals.
   - Values now have a `booleanValue` method which returns the boolean interpretation of the given value.
   - Built-in methods for which the return type is not fixed are now modeled as returning an unknown value by default.

## Changes to existing queries

| **Query**                  | **Expected impact**    | **Change**                                                       |
|----------------------------|------------------------|------------------------------------------------------------------|

> | Arbitrary file write during tarfile extraction (`py/tarslip`) | Fewer false negatives | Negations are now handled correctly in conditionals that may sanitize tainted values. |

Reviewer: "conditional" is an adjective, so suggest something like this, although I may well have picked the wrong term:
```suggestion
| Arbitrary file write during tarfile extraction (`py/tarslip`) | Fewer false negative results | Negations are now handled correctly in conditional expressions that may sanitize tainted values. |
```

Author: Ugh. I always forget to add the word "results". 
"Conditional expressions" sounds fine to me.

> | First parameter of a method is not named 'self' (`py/not-named-self`) | Fewer false positives | `__class_getitem__` is now recognized as a class method. |

Reviewer: ```suggestion
| First parameter of a method is not named 'self' (`py/not-named-self`) | Fewer false positive results | `__class_getitem__` is now recognized as a class method. |
```

> | Import of deprecated module (`py/import-deprecated-module) | Fewer false positives | Deprecated modules used for backwards compatibility are no longer reported.|

Reviewer: ```suggestion
| Import of deprecated module (`py/import-deprecated-module`) | Fewer false positive results | Deprecated modules that are used to provide backwards compatibility are no longer reported.|
```

> | Module imports itself (`py/import-own-module`) | Fewer false positives | Imports local to a given package are no longer classified as self-imports. |

Reviewer: ```suggestion
| Module imports itself (`py/import-own-module`) | Fewer false positive results | Imports local to a given package are no longer classified as self-imports. |
```
